### DISCUSSION TITLE
How to change the JVM flags being used in the startup script?

### DISCUSSION THREAD

**Discussion by rferreira** [jreleaser/jreleaser#1552] [Q&A] [answered]
Hello everyone, great tool, thank you tons. 

How do I change the JVM flags being used in the launcher script of the jlink assembler?

Apologies in advance if there's an obvious way to do this, I reviewed all the documentation and could not find it. 

Cheers!

**Comment by aalmiray** [comment]
I believe you may specify additional JVM flags by setting their values in a `JAVA_OPTS` environment variable.

**Reply by rferreira** [reply]
Thanks @aalmiray, can that be done as part of the environment/properties configuration like this? 

```
environment:
    properties:
        JVM_OPTS: -XX:+UseZGC -Xmx128m --enable-preview
````

**Reply by aalmiray** [reply]
I'm afraid not. Reading env vars from the model means the Java process has already started hence it's too late. You must set `JAVA_OPTS` as an env var in your executing shell.

**Reply by rferreira** [reply]
Not sure I follow, the JVM flags I need to pass are not for jreleaser itself, but instead, to my executable JAR. I see that launcher.tpl does support JVM_OPTS (which is great) - I just need to provide a default set of options first. I'll keep digging, thanks anyways.

I found this https://github.com/jreleaser/jreleaser/issues/1251 - but it doesn't appear to be supported in the jlink assembler - `java.options`  that is.

**Reply by aalmiray** [reply, marked as answer]
Oh, I misunderstood. I thought you meant the jreleaser CLI itself. For the time being there are only 2 launch script templates: java-archive and jlink. There’s no config option that will let you define configuration to be set as the starting values for `JAVA_OPTS` at the moment. I’ll raise an issue for this. 

However, you can supply your own template and use the `extraProperties` block in the jlink assembler to make this work. 

In the model:

```yaml
assemble:
  jlink:
    helloworld:
      active: ALWAYS
      imageName: '{{distributionName}}-{{projectVersion}}'
      extraProperties:
        jvmOpts: -XX:+UseZGC -Xmx128m --enable-preview
```

Next, generate the jlink launcher templates using the template command matching your distribution name

```
jreleaser `template generate --assembler jlink --distribution helloworld`
```

Then edit the templates to add the evaluated extra property `jlinkJvmOpts` (note the prefix)

**launcher.tpl**
```sh
JAVA_HOME=$APP_HOME
JAVACMD="$JAVA_HOME/bin/java"
JARSDIR="$APP_HOME/jars"
JAVA_OPTS="{{jlinkJvmOpts}}"
```

**launcher.bat.tpl**
```
set JAVA_HOME="%BASEDIR%"
set JAVACMD="%JAVA_HOME%\bin\java"
set JARSDIRS="%BASEDIR%\jars"
set JAVA_OPTS="{{jlinkJvmOpts}}"
```

Note that in this way any external values of `JAVA_OPTS` will be overridden. If you want to allow consumers to define additional values then you'd have to use a different variable name and combine both your values and any potential external values in the launcher script.

**Reply by aalmiray** [reply]
Filed https://github.com/jreleaser/jreleaser/issues/1553
